File: leetcode/must-do-medium-questions/lru-cache.rb
class ListNode
    attr_accessor :key, :val, :prev, :next

    def initialize(key, val)
        @key = key
        @val = val
    end
end

class LRUCache
#    :type capacity: Integer
    def initialize(capacity)
        @capacity = capacity
        @cache = {}
        @head = ListNode.new(-1, -1)
        @tail = ListNode.new(-1, -1)
        @head.next = @tail
        @tail.prev = @head
    end

#    :rtype: Integer
    def get(key)
        return -1 unless node = @cache[key]

        move_to_head(node)
        node.val
    end

#    :rtype: Void
    def put(key, value)
        if node = @cache[key]
            node.val = value
            move_to_head(node)
        else
            if @capacity.zero?
                @cache.delete pop_tail()
                @capacity += 1
            end
            node = ListNode.new(key, value)
            add_to_head(node)
            @cache[key] = node
            @capacity -= 1
        end
    end

    private

    def add_to_head(node)
        next_n = @head.next
        @head.next = node
        node.next = next_n
        next_n.prev = node
        node.prev = @head
    end

    def move_to_head(node)
        remove_node node
        add_to_head node
    end

    def remove_node(node)
        prev = node.prev
        next_n = node.next
        prev.next = next_n
        next_n.prev = prev
    end

    def pop_tail
        node = @tail.prev
        prev = node.prev
        remove_node node
        prev.next = @tail
        @tail.prev = prev
        node.key
    end
end

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache.new(capacity)
# param_1 = obj.get(key)
# obj.put(key, value)

File: leetcode/30-day-leetcoding-challenge-April/lru-cache.js

const ListNode = function(key, val) {
    this.key = key;
    this.val = val;
    this.prev = null;
    this.next = null;
}

/**
 * @param {number} capacity
 */
const LRUCache = function(capacity) {
    this.capacity = capacity;
    this.cache = new Map;
    this.head = new ListNode(-1, -1);
    this.tail = new ListNode(-1, -1);
    this.head.next = this.tail;
    this.tail.prev = this.head;
};

/** 
 * @param {number} key
 * @return {number}
 */
LRUCache.prototype.get = function(key) {
    const node = this.cache.get(key);
    if (!node) return -1;
    
    this.moveToHead(node);
    return node.val;
};

/** 
 * @param {number} key 
 * @param {number} value
 * @return {void}
 */
LRUCache.prototype.put = function(key, value) {
    let node = this.cache.get(key);
    if (node) {
        node.val = value;
        this.moveToHead(node);
    } else {
        if (this.capacity === 0) {
            const k = this.popTail();
            this.cache.set(k, undefined);
            this.capacity += 1;
        }
        node = new ListNode(key, value);
        this.addToHead(node);
        this.cache.set(key, node);
        this.capacity -= 1;
    }
};

LRUCache.prototype.addToHead = function(node) {
    const nextNode = this.head.next;
    this.head.next = node;
    node.next = nextNode;
    nextNode.prev = node;
    node.prev = this.head;
}

LRUCache.prototype.moveToHead = function(node) {
    this.removeNode(node);
    this.addToHead(node);
}
    
LRUCache.prototype.removeNode = function(node) {
    const prev = node.prev;
    const nextNode = node.next;
    prev.next = nextNode;
    nextNode.prev = prev;
}

LRUCache.prototype.popTail = function() {
    const node = this.tail.prev;
    this.prev = node.prev;
    this.removeNode(node);
    this.prev.next = this.tail;
    this.tail.prev = this.prev;
    return node.key;
}
/** 
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */
